zhusuan/sgmcmc.py

```diff
@@ -0,0 +1,546 @@
+            return a Tensor, representing the log joint likelihood of the
+            model. More conveniently, the user can also provide a
+            :class:`~zhusuan.framework.meta_bn.MetaBayesianNet` instance
+            instead of directly providing a log_joint function. Then a
+            log_joint function will be created so that `log_joint(obs) =
+            meta_bn.observe(**obs).log_joint()`.
+        :param observed: A dictionary of ``(string, Tensor)`` pairs. Mapping
+            from names of observed `StochasticTensor` s to their values
+        :param latent: A dictionary of ``(string, Variable)`` pairs.
+            Mapping from names of latent `StochasticTensor` s to corresponding
+            tensorflow `Variables` for storing their initial values and
+            samples.
+
+        :return: A default gradient function for sampling algorithm, whose
+            argument is a list of tensorflow Variables corresponding to the
+            latent variables.
+        :return: The same gradient function whose arguments are the `Variable`
+            list above **and** a `observed` dictionary, to allow users to
+            create a gradient function with dataset splitting in case of large
+            batch size and limited memory.
+        :return: The `Variable` list which will serve as the argument of the
+            gradient function.
+
+        .. note::
+
+            The return values are intended for enabling users to customize the
+            gradient function. If the user wants to use the default gradient
+            function, there is no need to fetch the return values (however this
+            function should always be called).
+
+        """
+        if callable(meta_bn):
+            self._meta_bn = None
+            self._log_joint = meta_bn
+        else:
+            self._meta_bn = meta_bn
+            self._log_joint = lambda obs: meta_bn.observe(**obs).log_joint()
+
+        self._observed = observed
+        self._latent = latent
+
+        latent_k, latent_v = [list(i) for i in zip(*six.iteritems(latent))]
+        for i, v in enumerate(latent_v):
+            if not isinstance(v, tf.Variable):
+                raise TypeError("latent['{}'] is not a tensorflow Variable."
+                                .format(latent_k[i]))
+
+        def _get_log_posterior(var_list, observed):
+            joint_obs = merge_dicts(dict(zip(latent_k, var_list)), observed)
+            return self._log_joint(joint_obs)
+
+        def _get_gradient(var_list, observed):
+            return tf.gradients(
+                _get_log_posterior(var_list, observed), var_list)
+
+        self._default_get_gradient = \
+            lambda var_list: _get_gradient(var_list, observed)
+        self._latent_k = latent_k
+        self._var_list = latent_v
+        return self._default_get_gradient, _get_gradient, latent_v
+
+    def sample(self, grad_func=None):
```

Comment:
Let's rename this to ``apply_updates`` and add another ``sample`` that does ``compute_grad_func`` as well as ``apply_updates``.